chromadb/api/async_fastapi.py

```diff
@@ -0,0 +1,634 @@
+
+
+class AsyncFastAPI(BaseHTTPClient, AsyncServerAPI):
+    # We make one client per event loop to avoid unexpected issues if a client
+    # is shared between event loops.
+    # For example, if a client is constructed in the main thread, then passed
+    # (or a returned Collection is passed) to a new thread, the client would
+    # normally throw an obscure asyncio error.
+    # Mixing asyncio and threading in this manner usually discouraged, but
+    # this gives a better user experience with practically no downsides.
+    # https://github.com/encode/httpx/issues/2058
+    _clients: Dict[int, httpx.AsyncClient] = {}
+
+    def __init__(self, system: System):
+        super().__init__(system)
+
+        system.settings.require("chroma_server_host")
+        system.settings.require("chroma_server_http_port")
+
+        self._opentelemetry_client = self.require(OpenTelemetryClient)
+        self._product_telemetry_client = self.require(ProductTelemetryClient)
+        self._settings = system.settings
+
+        self._api_url = AsyncFastAPI.resolve_url(
+            chroma_server_host=str(system.settings.chroma_server_host),
+            chroma_server_http_port=system.settings.chroma_server_http_port,
+            chroma_server_ssl_enabled=system.settings.chroma_server_ssl_enabled,
+            default_api_path=system.settings.chroma_server_api_default_path,
+        )
+
+    async def __aenter__(self) -> "AsyncFastAPI":
+        self._get_client()
+        return self
+
+    async def _cleanup(self) -> None:
+        while len(self._clients) > 0:
+            (_, client) = self._clients.popitem()
+            await client.aclose()
+
+    async def __aexit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        await self._cleanup()
+
+    @override
+    def stop(self) -> None:
+        super().stop()
+
+        @async_to_sync
+        async def sync_cleanup() -> None:
+            await self._cleanup()
+
+        sync_cleanup()
+
+    def _get_client(self) -> httpx.AsyncClient:
+        # Ideally this would use anyio to be compatible with both
+        # asyncio and trio, but anyio does not expose any way to identify
+        # the current event loop.
+        # We attempt to get the loop assuming the environment is asyncio, and
+        # otherwise gracefully fall back to using a singleton client.
+        loop_hash = None
+        try:
+            loop = asyncio.get_event_loop()
+            loop_hash = loop.__hash__()
+        except RuntimeError:
+            loop_hash = 0
+
+        if loop_hash not in self._clients:
+            self._clients[loop_hash] = httpx.AsyncClient(timeout=None)
+
+        return self._clients[loop_hash]
+
+    async def _make_request(
+        self, method: str, path: str, **kwargs: Dict[str, Any]
+    ) -> Any:
+        # Unlike requests, httpx does not automatically escape the path
+        escaped_path = urllib.parse.quote(path, safe="/", encoding=None, errors=None)
```

Comment:
should we port the sync code to httpx now (not necessarily in this PR) rather than leave this dangling?

Reply:
have a note about this in the OP :)--yes, I want to migrate the sync code soon as well